### ISSUE TITLE
Sequence.read fails for E.coli genbank record (0.4.0-dev)

### ISSUE THREAD

**Issue by ekopylova** [biocore/scikit-bio#1159]
Python exists with `Killed: 9` when attempting to load the E. coli GenBank record on local computer.
By rough calculation (thanks @RNAer ), there are ~260000 features in the record, and each feature is stored as a column of genome length (4686137) in positional_metadata. That will take 260000 \* 4686137 / 1024**3 = 1134GB.

file source: ftp://ftp.ncbi.nlm.nih.gov//genomes/Bacteria/Escherichia_coli_K_12_substr__DH10B_uid58979/NC_010473.gbk

```
(scikit-bio)hutgw-70-72-dhcp:jenya$ python
Python 2.7.10 (default, Jul 14 2015, 19:46:27) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.39)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from skbio import Sequence
>>> seq = Sequence.read("./NC_010473.gbk", format='genbank')
Killed: 9
```

**Comment by mortonjt** [comment]
Talking to @RNAer , it seems that the heart of the problem is the usage of dense pandas dataframes.
While it using sparse pandas dataframes may alleviate the problem, I'm not sure how scalable it is.

One alternative to this issue is using [Interval Trees](https://en.wikipedia.org/wiki/Interval_tree).  This allows for O(log n) access time to get all overlapping intervals.  Plus, it deals with the sparsity problem nicely, since it is only storing intervals.  One potential package to use is [bxpython](https://bitbucket.org/james_taylor/bx-python/wiki/Home).  This is personally what I use whenever I'm querying from genbank, gff and what-not.

**Comment by RNAer** [comment]
Thanks, @mortonjt . Interval Trees seems a very good option. It is not as convenient as pandas dataframe, but close. I haven't come up with a use case that can't be done with Interval Trees easily. Anyone else?

I think it all boils down the memory usage even with pandas `SparseDataFrame` and how big the genome we want to support?

**Comment by RNAer** [comment]
It should also solve the case (e) in #1137

**Comment by ebolyen** [comment]
@RNAer I'm not sure I see how an interval tree solves point (e). One thing we would need to think about is whether we could slice the sequence by an interval. Naively throwing an interval tree into the sequence object is unlikely to be successful in that regard. We would want an implementation that could return something like an index array or boolean vector which represented the response of a query or intersection. We would probably also want a mechanism to query membership in the tree using those same objects.

@shoyer you had mentioned attempting to implement interval trees in pandas at DS4DS last month. If you're not too busy, what did you have in mind exactly from an API perspective?

**Comment by ebolyen** [comment]
Also, interval trees need contiguous regions for it to work. Some features are unions of different regions.

**Comment by ebolyen** [comment]
Although I suppose if we considered a union to just be defined by the same instance everything would be ok, we would just have to do `k` separate queries for each sub-interval.

**Comment by shoyer** [comment]
@ebolyen Here is my half-finished `IntervalIndex` PR for pandas: https://github.com/pydata/pandas/pull/8707

It includes a self-contained `IntervalTree` written in Cython, which could also be extracted separately if desired.

From an API perspective, the plan would be for `IntervalIndex` to support several types of queries:
- `get_loc`: this pulls out all locations in the array of intervals that overlap with the given point or interval
- `get_indexer`: this does the same, but looks up an entire list of points or intervals all at once.

These methods correspond to those called by the `.loc` method on Series/DataFrame objects.

We could also pretty easily add a method for looking up all intervals *contained* inside a given interval.

**Comment by ebolyen** [comment]
Thanks @shoyer! Out of curiosity how would the `IntervalIndex` be used in a `Series` or `DataFrame`? 

If I understand the linked code correctly (which is unlikely): 

For our use case we could imagine building a `Series` with the `IntervalIndex` and `np.arange(n)` for the data. Then we would be able to associate labels with intervals. Would slicing the series with `.loc` then return only those "cells" in the interval, effectively producing an index-array?

**Comment by shoyer** [comment]
@ebolyen Yes, that would work, though you might also put your data directly in a DataFrame or Series instead of creating a dummy index array with `np.arange`.

**Comment by ebolyen** [comment]
:+1: I was imagining slicing our object, so adjusting, I could see something like this:

`dna[dna.positional_metadata['dummy_index'].loc['feature 1']]` 
^ this should probably be a helper: `dna[dna.feature_location('feature 1')]`
and
`dna.positional_metadata['quality'].loc['feature 1']` 

Would it be possible to support multiple intervals per label?

**Comment by shoyer** [comment]
> Would it be possible to support multiple intervals per label?

I don't think this would belong in pandas directly -- that's not how pandas Index works. But you could pretty easily have a proxy object (e.g., an integer pandas.Series with a pandas.IntervalIndex index) that maps multiple intervals to the same integer label.

**Comment by RNAer** [comment]
@ebolyen , I see your point we need to write/maintain extra code wrt the interval approach, but in terms of API usage, I don't see how it cannot be designed to look like the current `positional_metadata` implementation?

Anyway, we need a solution. Certainly it would be best for `pandas` to have this functionality (Thanks, @shoyer !). It is now blocking @ekopylova 's HGT project, the development of micronota, and other usage of scikit-bio. Do we have an ETA on this functionality in `pandas`? Is there anything we can help to accelerate it? @mortonjt is willing to help if needed.

**Comment by gregcaporaso** [comment]
@RNAer, at this point it's not clear what the feature request is, so we'd need to sort that out before coming up with a solution and an ETA. Are you guys using *all* of the metadata in that record? What features specifically do you need to access and operate on from that record? It'd probably be easiest to discuss this, and how we want to support it, on a call. Note that the focus in my lab *needs* to be on the items we discussed on last Friday's call through mid-December for us to get the 0.4.1 release out. That's our highest priority at the moment.

**Comment by ekopylova** [comment]
For simulating horizontal gene transfers, I need the CDS features for two microbial genomes (donor & recipient) at a time. Previously I was using FeatureExtract, a python script for parsing GenBank records.

**Comment by RNAer** [comment]
@gregcaporaso , totally understand. We were just saying the current `positional_metadata` blew up the mem usage to ~1TB to read a E.coli GenBank file of 15Mb, which is obviously difficult to use. We can see the roaring mem footprint during feature parsing for a GenBank file reduced to have only ~20 features.
![screen shot 2015-10-30 at 10 20 56 pm](https://cloud.githubusercontent.com/assets/1000056/10862006/847c538c-7f54-11e5-88aa-7b537b4954ec.png)

I just tested with `SparseSeries` and `SparseDataFrame`. It seems not helping.

The solutions are:
1). waiting the `pandas` to have Stephan's interval tree.

2). implementing interval tree with or without bxpython

3). temporarily re-design GenBank interface to read only selected features. But we do want to keep all the features in memory in a lot of cases.

Personally, I would go with option 1) & 2), and we need an ETA for 1)

**Comment by ebolyen** [comment]
@RNAer when you tested it with the SparseSeries did you set the `fill_value` to `False`? By default it is `NaN` and so you wouldn't see any difference.

**Comment by RNAer** [comment]
I used it. It reduced only by half:
<img width="1206" alt="screen shot 2015-10-31 at 1 54 14 am" src="https://cloud.githubusercontent.com/assets/1000056/10862748/0a1b961a-7f73-11e5-85b6-fd07fe651aa8.png">


The code is here: https://github.com/RNAer/scikit-bio/commit/c00bb7cb4b93619405c78d7482f7f4d2671a99ec

**Comment by RNAer** [comment]
To be fair, the GenBank record without any features takes 100Mb, so the `SparseDataFrame` reduces the mem usage from 900Mb to 300Mb. 3 folds differences.

**Comment by ebolyen** [comment]
@RNAer Thanks for posting these benchmarks! The 3 fold difference is certainly nice, but I think you are right that it isn't going to solve this problem fundamentally. It would be interesting to see how an interval tree compares. @RNAer would you mind posting your test file and script? (Where does the nice line-chart come from?) I might try playing with @shoyer's `IntervalIndex` assuming the merge conflicts aren't too bad.

**Comment by RNAer** [comment]
Thanks, @ebolyen ! [test file](https://drive.google.com/file/d/0B6wW1JWx55lRNFNWTkwwVmJlTDA/view?usp=sharing) and [script](https://drive.google.com/file/d/0B6wW1JWx55lRLW1xWG5CLTQ4Q2M/view?usp=sharing). The mem profiling is done with python package memory_profiler. The command is:

     mprof run python /tmp/test.py /tmp/NC_010473.gbk && mprof plot -o /tmp/f.pdf

**Comment by jairideout** [comment]
@RNAer thanks for looking into this! @ebolyen and I will start investigating the memory issues today -- it is concerning to see that a GenBank sequence *without* any features requires ~100MB when the file is ~5MB.

**Comment by shoyer** [comment]
I spent a few hours looking at my IntervalIndex code again last night and fixing stuff. It's actually closer to being ready than I thought before -- you can see the checklist in the PR. Most of the stuff to do at this point is integration stuff and more exhaustive tests. I don't think there are any major gaps in core functionality.

I did stall out when I attempted to rebase -- I need to rewrite the Cython header for `Int64Vector`, which was updated not too long ago so it could release the GIL. If any of you are experienced with that sort of thing, guidance would be appreciated (I'm mostly bumbling around in the dark there). If not, no worries, I'm sure @jreback can help.

Once I get the rebase fixed, it would certainly be helpful if some of you could give it a try and report back on your success. Also, it would be great if I could get a technical code review on the IntervalTree part of the PR, which is mostly self-contained, though it's currently not well documented. That would also serve you well if you choose to implement your own version. Let me know if any of you are interested in this and I'll update the code with some notes on my implementation.

Finally, I can't make promises on delivery timeline. If everything went well, this could be ready to merge in a few weeks, but pandas just had a major release so we won't have another one for another month or two.

**Comment by gregcaporaso** [comment]
@RNAer, @ekopylova I think we'll need to handle this as follows. @jairideout and @ebolyen are going to look into the memory issue in loading genbank files *without* features, as that sounds like a memory bug that is separate from the positional metadata issue described here. They're not going to be able to work on the positional metadata sparsity/interval tree issue before [0.4.1](https://github.com/biocore/scikit-bio/milestones/0.4.1) (scheduled for early December). We just don't have the developer bandwidth to spend on this (which we think will be a substantial development effort) and meet that milestone target. 

What I would suggest is that if you need this now, that you start developing it in your fork of scikit-bio. Since we're working on the ``alignment`` and ``distance`` subpackages primarily for 0.4.1, we shouldn't run into any challenging merge conflicts. This would let you guys explore different solutions and get something that works to address your immediate needs. After 0.4.1, we can work with you to get the best solution integrated in scikit-bio. The best way to get input on a potential solution (to avoid a bunch of suggestions for changes after you've done a lot of work) would be to [create an RFC](https://github.com/biocore/scikit-bio-rfcs) for this. This worked well for ``TablularMSA``, in that we were able to get the design solidified before spending a lot of time coding. (Note that this approach is similar to what we discussed for QIIME 2 development: if we need something urgently for QIIME that isn't in scikit-bio yet, we can develop in QIIME, and port it to scikit-bio as we stabilize the API.)

Alternatively, it sounds like you had a non-sckit-bio option working. If it's urgent and you don't have time to contribute to scikit-bio in this way right now, going with that solution (unfortunately) might be the best approach.

**Comment by jairideout** [comment]
Just an update on the potential memory bug @ebolyen and I were looking into:

Memory usage appears to be what we'd expect: importing scikit-bio alone takes ~74MB and loading the sequence is another ~46MB, of which `positional_metadata` is ~35MB. The 100MB for loading a sequence without features isn't so surprising anymore.

A slight optimization would be to put the "source" feature's metadata in `metadata` since it applies to the entire sequence anyways. This would avoid always having a DataFrame with at least one column of `True`.

**Comment by shoyer** [comment]
FYI, the pandas `IntervalIndex` should be successfully rebased now if any of you want to try it out.

**Comment by ebolyen** [comment]
Thanks @shoyer!

**Comment by shoyer** [comment]
Also, to be clear -- exactly what sort of data do you want to store in the intervals? My `IntervalTree` implementation currently converts everything to `float64`, but probably could be adapted to use fused types if necessary.

**Comment by RNAer** [comment]
@gregcaporaso , sounds good. Thanks for your suggestion.

@jairideout , surprised to see importing alone takes 74MB. It makes sense to move "source" to `metadata`

@shoyer , thanks! I believe we only need `int` type.

**Comment by mortonjt** [comment]
Just as a heads up, I've been testing out @shoyer 's PR.  It definitely seems functional.
Attaching a snippet of code using a gff file of the E coli genome below.

ftp://ftp.ncbi.nlm.nih.gov/genomes/genbank/bacteria/Escherichia_coli/reference/GCA_000005845.2_ASM584v2

```python
>>> import numpy as np
>>> import pandas as pd
>>> from pandas.core.interval import Interval, IntervalIndex
>>> from pandas.core.index import Index
>>> from pandas.lib import IntervalTree


>>> gff = pd.read_table('GCA_000005845.2_ASM584v2_genomic.gff', skiprows=7, header=None,
...                     names=['acc', 'region', 'start', 'end', '1', 'strand', '2', 'meta'],
...                     comment='#')
>>> intervals = list(map(tuple, gff.loc[:, ['start', 'end']].values))
>>> index = IntervalIndex.from_tuples(intervals)
>>> df = pd.DataFrame(gff.values, index=index,
                      columns=['acc', 'region', 'start', 'end',
                               '1', 'strand', '2', 'meta'])
>>> df.loc[4640402, :]
                  acc  region start      end  1 strand  2  \
(1, 4641652]  Genbank  region     1  4641652  .      +  .   

                                                           meta  
(1, 4641652]  ID=id0;Dbxref=taxon:511145;Is_circular=true;Na...  

>>> df.loc[Interval(4640541, 4640944), :]

                        acc  region    start      end  1 strand  2  \
(4640942, 4641628]  Genbank    gene  4640942  4641628  .      +  .   
(4640942, 4641628]  Genbank     CDS  4640942  4641628  .      +  0   
(4640402, 4640542]  Genbank     CDS  4640402  4640542  .      +  0   
(4640402, 4640542]  Genbank    gene  4640402  4640542  .      +  .   
(1, 4641652]        Genbank  region        1  4641652  .      +  .   

                                                                 meta  
(4640942, 4641628]  ID=gene4497;Dbxref=EcoGene:EG12309;Name=yjtD;g...  
(4640942, 4641628]  ID=cds4318;Parent=gene4497;Dbxref=ASAP:ABE-001...  
(4640402, 4640542]  ID=cds4317;Parent=gene4496;Dbxref=ASAP:ABE-001...  
(4640402, 4640542]  ID=gene4496;Dbxref=EcoGene:EG12601;Name=yjjY;g...  
(1, 4641652]        ID=id0;Dbxref=taxon:511145;Is_circular=true;Na...  
```

So, what's left to discuss on this?  This new index seems to be appropriate for the positional_metadata.

**Comment by RNAer** [comment]
Just keep a record here. When Genbank parser uses the new `IntervalMetadata`, the same test file of gbk file takes ~110mb:

![screen shot 2016-10-04 at 1 16 02 pm](https://cloud.githubusercontent.com/assets/1000056/19090761/c259c194-8a34-11e6-8a2a-744e6ffab78a.png)

And the [full E coli genbank record](https://www.ncbi.nlm.nih.gov/nuccore/U00096.3#sequence_545778205) with 9411 anntations takes ~150mb:

![screen shot 2016-10-04 at 1 23 59 pm](https://cloud.githubusercontent.com/assets/1000056/19091048/d8a4cae2-8a35-11e6-965a-67fa2d075cad.png)

**Comment by mortonjt** [comment]
Woo!

**Comment by jairideout** [comment]
Fantastic!!!

**Comment by jairideout** [comment]
Fixed in #1440.
